funnel/views/email_events.py

```diff
@@ -0,0 +1,119 @@
+from flask import request
+
+import requests
+
+from coaster.views import render_with
+
+from .. import app
+from ..models import EmailAddress, db
+from ..transports.email.aws_ses import (
+    Bounce,
+    Complaint,
+    Delivery,
+    DeliveryDelay,
+    SesEvent,
+    SesProcessorAbc,
+    SnsNotificationType,
+    SnsValidator,
+    SnsValidatorException,
+)
+
+
+class SesProcessor(SesProcessorAbc):
+    """SES message processor."""
+
+    # `EmailAddress.add` does an implicit `.get`, but we call `.get` first because
+    # `.add` will fail if the address is blocked, while `.get` won't. Why add if we've
+    # never seen this email address before? Because it may have originated in Hasjob
+    # or elsewhere in shared infrastructure.
+
+    def bounce(self, bounce: Bounce) -> None:
+        for bounced in bounce.bounced_recipients:
+            email_address = EmailAddress.get(bounced.email)
+            if not email_address:
+                email_address = EmailAddress.add(bounced.email)
+            if bounce.is_hard_bounce:
+                email_address.mark_hard_fail()
+            else:
+                email_address.mark_soft_fail()
+
+    def delayed(self, delayed: DeliveryDelay) -> None:
+        for failed in delayed.delayed_recipients:
+            email_address = EmailAddress.get(failed.email)
+            if not email_address:
+                email_address = EmailAddress.add(failed.email)
+            email_address.mark_soft_fail()
+
+    def complaint(self, complaint: Complaint) -> None:
+        for complained in complaint.complained_recipients:
+            if complaint.complaint_feedback_type == "not-spam":
+                email_address = EmailAddress.get(complained.email)
+                if not email_address:
+                    email_address = EmailAddress.add(complained.email)
+                email_address.mark_active()
+            else:
+                EmailAddress.mark_blocked(complained.email)
+
+    def delivered(self, delivery: Delivery) -> None:
+        # TODO: Why is this a string and not a structure? Is this correct?
+        for sent in delivery.recipients:
+            email_address = EmailAddress.get(sent)
+            if not email_address:
+                email_address = EmailAddress.add(sent)
+            email_address.mark_sent()
```

Comment:
This event has a different data structure from the others. Appears to be accidental.

Reply:
Confirmed from documentation that this is correct.